### PR TITLE
Add Dice Coefficient

### LINKED ISSUE #368: Adding Dice Coefficient Metric for Image Segmentation
I wanted to implement Dice Coefficient as a metric for image segmentation problems. any interest?

### PR DESCRIPTION
**Pull request by PyExtreme**
Fixes #368

Description: Adding Dice Coefficient Metric for Image Segmentation

Hi @vfdev-5 , this is just an initial work on Dice Coefficient. I will be adding more to it and will be adding the test as well.  

Thank you

### CODE REVIEW

**Comment by PyExtreme** [comment]
@vfdev-5 , could you please suggest me what should be logic for the test? I am a bit confused regarding it.

File: ignite/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -238,3 +238,22 @@ def cmRecall(cm, average=True):
...
+    if not isinstance(cm, ConfusionMatrix):
+        raise TypeError("Argument cm should be instance of ConfusionMatrix, but given {}".format(type(cm)))
+
+    if ignore_index is not None:
+        if not (isinstance(ignore_index, numbers.Integral) and 0 <= ignore_index < cm.num_classes):
+            raise ValueError("ignore_index should be non-negative integer, but given {}".format(ignore_index))
+
+    tp = cm.confusion_matrix.diag().sum()
+    fp_plus_fn = cm.confusion_matrix.sum() - tp
+    return (2.0 * tp) / (2.0 * tp + fp_plus_fn)
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]:
```suggestion
    return (2.0 * tp) / (2.0 * tp + fp_plus_fn + 1e-15)
```

File: ignite/metrics/confusion_matrix.py [hunk 2]
```diff
@@ -238,3 +238,22 @@ def cmRecall(cm, average=True):
     if average:
         return recall.mean()
     return recall
+
+
+def dice_coefficient(cm, ignore_index=None):
+    """Calculates Dice Coefficient for a given Confusion Matrix.
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]:
```suggestion
    """Calculates Dice Coefficient for a given Confusion Matrix.

```

**Comment by vfdev-5** [comment]
Concerning tests, please take a look at the test of IoU : 
https://github.com/pytorch/ignite/blob/master/tests/ignite/metrics/test_confusion_matrix.py#L301

File: ignite/metrics/confusion_matrix.py [hunk 3]
```diff
@@ -238,3 +238,24 @@ def cmRecall(cm, average=True):
     if average:
         return recall.mean()
     return recall
+
+
+def dice_coefficient(cm, ignore_index=None):
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]: Could you please change the name to `DiceCoefficient`

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 4]
```diff
@@ -7,8 +7,8 @@
 
 from ignite.exceptions import NotComputableError
 from ignite.metrics import ConfusionMatrix, IoU, mIoU
-from ignite.metrics.confusion_matrix import cmAccuracy, cmPrecision, cmRecall
-
+from ignite.metrics.confusion_matrix import cmAccuracy, cmPrecision, cmRecall, dice_coefficient
+import random
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py]: Maybe we can use `torch.rand` instead of `random`?

Comment (PyExtreme) [reply tests/ignite/metrics/test_confusion_matrix.py]: Hi @vfdev-5 , since we are using numpy array, I have used random here.

Comment (vfdev-5) [reply tests/ignite/metrics/test_confusion_matrix.py]: In this case you can directly `np.random`

File: ignite/metrics/confusion_matrix.py [hunk 5]
```diff
@@ -238,3 +238,24 @@ def cmRecall(cm, average=True):
...
+    """
+
+    if not isinstance(cm, ConfusionMatrix):
+        raise TypeError("Argument cm should be instance of ConfusionMatrix, but given {}".format(type(cm)))
+
+    if ignore_index is not None:
+        if not (isinstance(ignore_index, numbers.Integral) and 0 <= ignore_index < cm.num_classes):
+            raise ValueError("ignore_index should be non-negative integer, but given {}".format(ignore_index))
+
+    eps = 1e-15
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]: Still missing logic with `ignore_index`

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 6]
```diff
@@ -577,3 +577,39 @@ def test_multinode_distrib_cpu(distributed_context_multi_node_gloo):
 def test_multinode_distrib_gpu(distributed_context_multi_node_nccl):
     device = "cuda:{}".format(distributed_context_multi_node_nccl['local_rank'])
     _test_distrib_multiclass_images(device)
+
+
+def test_dice_coefficient(ignore_index=None):
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py]: Why `ignore_index=None` as argument here ?

File: ignite/metrics/confusion_matrix.py (hunk 5 above)
Comment (PyExtreme) [reply ignite/metrics/confusion_matrix.py]: Sure, @vfdev-5 , will make these changes

**Comment by vfdev-5** [comment]
@PyExtreme graceful ping

**Comment by PyExtreme** [comment]
> @PyExtreme graceful ping

Hi @vfdev-5 , I had been on vacation so could not work. Now, I am free.

**Comment by PyExtreme** [comment]
Hi @vfdev-5 , I have added the logic for _`ignore_index`_.

Please have a look at the latest commit.

File: ignite/metrics/confusion_matrix.py [hunk 7]
```diff
@@ -238,3 +238,27 @@ def cmRecall(cm, average=True):
...
+        ignore_index (int, optional): index to ignore, e.g. background index
+    """
+
+    if not isinstance(cm, ConfusionMatrix):
+        raise TypeError("Argument cm should be instance of ConfusionMatrix, but given {}".format(type(cm)))
+
+    if ignore_index is not None:
+        if not (isinstance(ignore_index, numbers.Integral) and 0 <= ignore_index < cm.num_classes):
+            raise ValueError("ignore_index should be non-negative integer, but given {}".format(ignore_index))
+        cm.confusion_matrix[indices] = 0
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]: `indices` are not defined

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 8]
```diff
@@ -577,3 +577,39 @@ def test_multinode_distrib_cpu(distributed_context_multi_node_gloo):
 def test_multinode_distrib_gpu(distributed_context_multi_node_nccl):
     device = "cuda:{}".format(distributed_context_multi_node_nccl['local_rank'])
     _test_distrib_multiclass_images(device)
+
+
+def test_dice_coefficient():
+    n_classes = 3
+    logits = [[random.random() for j in range(n_classes)] for i in range(5)]
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py]: replace `random.random` by `np.random`

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 9]
```diff
@@ -577,3 +577,39 @@ def test_multinode_distrib_cpu(distributed_context_multi_node_gloo):
 def test_multinode_distrib_gpu(distributed_context_multi_node_nccl):
     device = "cuda:{}".format(distributed_context_multi_node_nccl['local_rank'])
     _test_distrib_multiclass_images(device)
+
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py]: Add a test with `ignore_index`

**Comment by PyExtreme** [comment]
@vfdev-5 , sure will make the required changes.

**Comment by vfdev-5** [comment]
@PyExtreme any updates ?

**Comment by PyExtreme** [comment]
> @PyExtreme any updates

Hi @vfdev-5 , I am a bit confused regarding the logic for ignore_index in test file. Do you have any ideas?

**Comment by vfdev-5** [comment]
@PyExtreme it is the same logic as for IoU. Please, take a look at the code to understand how it is done. Thanks

**Comment by PyExtreme** [comment]
> @PyExtreme it is the same logic as for IoU. Please, take a look at the code to understand how it is done. Thanks

Got it. Will finish it this week. Sorry for the delay, actually, I have been quite occupied with some work.

**Comment by PyExtreme** [comment]
Hi @vfdev-5 , I made and pushed few changes day before yesterday. Please have a look.

File: tests/ignite/metrics/test_confusion_matrix.py [hunk 10]
```diff
@@ -577,3 +577,39 @@ def test_multinode_distrib_cpu(distributed_context_multi_node_gloo):
...
+
+def test_dice_coefficient():
+    n_classes = 3
+    logits = [[np.random.random() for j in range(n_classes)] for i in range(5)]
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_confusion_matrix.py]: @PyExtreme please create directly 2D arrays as `np.random.rand(5, n_classes)` for logits and y_true

File: ignite/metrics/confusion_matrix.py [hunk 11]
```diff
@@ -238,3 +238,27 @@ def cmRecall(cm, average=True):
...
+        ignore_index (int, optional): index to ignore, e.g. background index
+    """
+
+    if not isinstance(cm, ConfusionMatrix):
+        raise TypeError("Argument cm should be instance of ConfusionMatrix, but given {}".format(type(cm)))
+
+    if ignore_index is not None:
+        if not (isinstance(ignore_index, numbers.Integral) and 0 <= ignore_index < cm.num_classes):
+            raise ValueError("ignore_index should be non-negative integer, but given {}".format(ignore_index))
+        cm.confusion_matrix[ignore_index] = 0
```
Comment (vfdev-5) [inline ignite/metrics/confusion_matrix.py]: I think this is a not good idea to modify `confusion_matrix` as it can be, for example, also used in other metrics which use this info.
